File: langs/typescript/src/viewmodels/app-vm.ts
/**
 * AppVM — root ViewModel and app-shell observables.
 *
 * Owns observable state that lives above any single scenario: the active
 * theme, the runtime mode, and the child ScenarioVM. Every View binds to
 * AppVM and navigates down to AppVM.scenario for scenario-specific state.
 *
 * Contract:
 *   - theme: string observable (canonical 'dark' + 'light' must work in
 *     all impls; impls may extend with framework-specific theme names).
 *   - mode: 'web' | 'native' | 'tauri', immutable after construction.
 *   - scenario: ScenarioVM, constant reference (only its internal state
 *     mutates; AppVM.scenario itself is never reassigned).
 *
 * Persistence: theme round-trips through localStorage under
 * "guidearch.theme". Missing / unreadable / malformed value silently
 * falls back to 'dark' and the next setTheme rewrites it.
 *
 * setTheme(name) validates against KNOWN_THEMES; an unknown name appends
 * a warning and leaves theme unchanged (never throws — the View's theme
 * picker should never blow up the app).
 */

import { ComponentVMOf, RelayCommandOf, MessageHub, NullDispatcher, type IMessageHub } from 'vmx';

import { makeScenarioVm, type ScenarioVM } from './scenario-vm.js';

// ---------------------------------------------------------------------------
// Theme set
// ---------------------------------------------------------------------------
// 'dark' and 'light' are MANDATED by spec across all impls. Additional
// framework-supported themes (e.g. high-contrast variants) can be added
// by calling registerTheme(name) at startup before the first AppVM constructs.
export const KNOWN_THEMES: Set<string> = new Set(['dark', 'light']);

export const DEFAULT_THEME = 'dark';

/**
 * Register an additional theme name (idempotent). Mirrors Python's
 * `register_theme(name)` and C#'s `AppVMFactory.RegisterTheme(name)`.
 */
export function registerTheme(name: string): void {
  KNOWN_THEMES.add(name);
}

// ---------------------------------------------------------------------------
// Persistence — small JSON-ish wrapper around localStorage
// ---------------------------------------------------------------------------
const STORAGE_KEY = 'guidearch.theme';

function loadPersistedTheme(): string {
  // Guard for SSR / Node-test envs where localStorage doesn't exist.
  if (typeof localStorage === 'undefined') return DEFAULT_THEME;
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (raw && KNOWN_THEMES.has(raw)) return raw;
  } catch {
    // Quota exceeded, security policy, etc. — silently fall through.
  }
  return DEFAULT_THEME;
}

function persistTheme(theme: string): void {
  if (typeof localStorage === 'undefined') return;
  try {
    localStorage.setItem(STORAGE_KEY, theme);
  } catch {
    // Storage failure is non-fatal — the live VM state wins, and the next
    // launch quietly falls back to default.
  }
}

// ---------------------------------------------------------------------------
// Mode detection
// ---------------------------------------------------------------------------
export type AppMode = 'web' | 'native' | 'tauri';

function detectMode(): AppMode {
  // Tauri runtime exposes the __TAURI__ global on the window object. Node
  // tests (vitest) don't have window. Anything else with a window is plain
  // browser/web mode.
  if (typeof window === 'undefined') return 'web';
  if (typeof (window as Window & { __TAURI__?: unknown }).__TAURI__ !== 'undefined') {
    return 'tauri';
  }
  return 'web';
}

// ---------------------------------------------------------------------------
// State shape
// ---------------------------------------------------------------------------
export interface AppState {
  theme: string;
  mode: AppMode;
  warnings: readonly string[];
}

// ---------------------------------------------------------------------------
// VM type
// ---------------------------------------------------------------------------
export type AppVM = ComponentVMOf<AppState> & {
  readonly setThemeCmd: RelayCommandOf<string>;
  readonly scenario: ScenarioVM;
  readonly hub: IMessageHub;

  /** Convenience — equivalent to setThemeCmd.execute(name). */
  setTheme(name: string): void;
};

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------
export interface MakeAppVmOptions {
  /**
   * Inject a different theme set (testing or impl extension). When omitted,
   * the module-level KNOWN_THEMES is used.
   */
  knownThemes?: Set<string>;
  /**
   * Inject a custom load function (testing). When omitted, reads from
   * localStorage.
   */
  loadTheme?: () => string;
  /**
   * Inject a custom persist function (testing). When omitted, writes to
   * localStorage.
   */
  persistTheme?: (theme: string) => void;
  /**
   * Force a specific mode (testing). When omitted, detected from runtime.
   */
  mode?: AppMode;
  /**
   * Reuse an existing ScenarioVM (testing). When omitted, a fresh one is
   * constructed via makeScenarioVm().
   */
  scenario?: ScenarioVM;
  /**
   * Inject an explicit MessageHub (testing). When omitted, a fresh one is
   * constructed internally — and exposed as AppVM.hub either way.
   */
  hub?: MessageHub;
}

export function makeAppVm(opts: MakeAppVmOptions = {}): AppVM {
  const known = opts.knownThemes ?? KNOWN_THEMES;
  const load = opts.loadTheme ?? loadPersistedTheme;
  const save = opts.persistTheme ?? persistTheme;
  const mode = opts.mode ?? detectMode();
  const scenario = opts.scenario ?? makeScenarioVm();

  const hub = opts.hub ?? new MessageHub();

  let _vm: ComponentVMOf<AppState> | null = null;

  const initialTheme = load();
  // If load returned something the canonical KNOWN_THEMES doesn't recognise
  // (e.g. an impl extension wasn't registered this launch), fall back
  // silently — matches the "missing/malformed → default" contract.
  const startTheme = known.has(initialTheme) ? initialTheme : DEFAULT_THEME;

  function _getModel(): AppState {
    if (_vm === null) {
      return { theme: startTheme, mode, warnings: [] };
    }
    return _vm.model;
  }

  function _setState(patch: Partial<AppState>): void {
    if (_vm === null) return;
    _vm.model = { ..._vm.model, ...patch };
  }

  function setThemeImpl(name: string): void {
    if (!known.has(name)) {
      const msg = `Unknown theme: ${name}`;
      const warnings = [..._getModel().warnings, msg];
      _setState({ warnings });
      return;
    }
    if (name === _getModel().theme) return; // no-op, don't refire
    _setState({ theme: name });
    save(name);
  }

  const setThemeCmd = RelayCommandOf.builder<string>().task(setThemeImpl).build();

  const vm = ComponentVMOf.builder<AppState>()
    .name('app-vm')
    .model({ theme: startTheme, mode, warnings: [] })
    .services(hub, NullDispatcher.INSTANCE)
    .modeledHinter((m) => `theme=${m.theme} mode=${m.mode}`)
    .build();

  vm.construct();
  _vm = vm;

  Object.defineProperties(vm, {
    setThemeCmd: {
      value: setThemeCmd,
      writable: false,
      enumerable: true,
      configurable: false,
    },
    scenario: {
      value: scenario,
      writable: false,
      enumerable: true,
      configurable: false,
    },
    hub: {
      value: hub,
      writable: false,
      enumerable: true,
      configurable: false,
    },
    setTheme: {
      value: setThemeImpl,
      writable: false,
      enumerable: true,
      configurable: false,
    },
  });

  return vm as AppVM;
}

File: langs/python/src/guidearch/viewmodels/app_vm.py
"""AppVM — root ViewModel for the GuideArch app shell.

Owns observable state that lives above any single scenario: the active
theme, the runtime mode, and the child :class:`ScenarioVM`. Every View
binds to AppVM and navigates down to ``app_vm.scenario`` for scenario-
specific state.

Cross-impl contract
-------------------
* ``theme`` — open string observable. "dark" and "light" are mandated
  in all three impls; impls may extend by calling :func:`register_theme`
  at startup (before the first AppVM is constructed) to expose
  framework-supported variants.
* ``mode`` — ``"web" | "native" | "tauri"``. Set at construction and
  immutable on the public surface. There is no ``set_mode`` command.
* ``scenario`` — :class:`ScenarioVM`. Reference is stable; only its
  internal state mutates.

Persistence
-----------
Theme round-trips through
``platformdirs.user_config_dir("guidearch") / "settings.json"`` (atomic
write — ``.tmp`` then rename). Missing / unreadable / unknown values
silently fall back to :data:`DEFAULT_THEME` and the next successful
:meth:`AppVM.set_theme` rewrites the file.

``set_theme`` validates against the registered theme set. An unknown
name appends a warning and leaves state unchanged — it never raises.
The View's theme picker should be safe to feed arbitrary user input.
"""

from __future__ import annotations

import json
from collections.abc import Callable
from pathlib import Path
from typing import Literal

import reactivex as rx
from platformdirs import user_config_dir
from reactivex.subject import Subject
from vmx.commands.relay_command import RelayCommandOf
from vmx.messages.property_changed import PropertyChangedMessage
from vmx.messages.protocols import Message
from vmx.services.message_hub import MessageHub

from guidearch.viewmodels.scenario_vm import ScenarioVM

# ---------------------------------------------------------------------------
# Theme registry
# ---------------------------------------------------------------------------

DEFAULT_THEME = "dark"

# "dark" and "light" are mandated cross-impl. Impls extend at startup
# via register_theme() before constructing the first AppVM.
_known_themes: set[str] = {"dark", "light"}


def register_theme(name: str) -> None:
    """Register an additional theme name (idempotent)."""
    _known_themes.add(name)


def known_themes() -> frozenset[str]:
    """Snapshot of currently-registered theme names."""
    return frozenset(_known_themes)


def is_known_theme(name: str) -> bool:
    return name in _known_themes


# ---------------------------------------------------------------------------
# Persistence
# ---------------------------------------------------------------------------

Mode = Literal["web", "native", "tauri"]


def _settings_path() -> Path:
    return Path(user_config_dir("guidearch")) / "settings.json"


def _load_persisted_theme() -> str:
    path = _settings_path()
    try:
        if not path.exists():
            return DEFAULT_THEME
        with path.open("r", encoding="utf-8") as fh:
            data = json.load(fh)
        if isinstance(data, dict):
            name = data.get("theme")
            if isinstance(name, str) and name in _known_themes:
                return name
    except (OSError, json.JSONDecodeError):
        # I/O, permission, malformed JSON — all non-fatal; fall through.
        pass
    return DEFAULT_THEME


def _persist_theme(theme: str) -> None:
    path = _settings_path()
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        tmp = path.with_suffix(path.suffix + ".tmp")
        with tmp.open("w", encoding="utf-8") as fh:
            json.dump({"theme": theme}, fh)
        tmp.replace(path)
    except OSError:
        # Best-effort: in-memory state is always authoritative.
        pass


# ---------------------------------------------------------------------------
# AppVM
# ---------------------------------------------------------------------------


class AppVM:
    """Root ViewModel; mirrors the TS ``AppVM`` and the C# ``ComponentVM<AppState>``.

    Construction
    ------------
    Use the :func:`make_app_vm` factory. Tests may inject ``load_theme`` /
    ``persist_theme`` stubs to round-trip through a fake store, and
    ``scenario`` to reuse an existing :class:`ScenarioVM`.
    """

    def __init__(
        self,
        *,
        mode: Mode = "native",
        load_theme: Callable[[], str] | None = None,
        persist_theme: Callable[[str], None] | None = None,
        scenario: ScenarioVM | None = None,
        hub: MessageHub[Message] | None = None,
    ) -> None:
        self._load = load_theme or _load_persisted_theme
        self._save = persist_theme or _persist_theme
        self._scenario = scenario or ScenarioVM()
        self.hub: MessageHub[Message] = hub or MessageHub()

        start = self._load()
        if start not in _known_themes:
            start = DEFAULT_THEME

        # State
        self._theme: str = start
        self._mode: Mode = mode
        self._warnings: tuple[str, ...] = ()

        # property_changed observable (matches ScenarioVM pattern)
        self._property_changed_subject: Subject[str] = Subject()

        # set_theme as a fluent command for parity with the other impls.
        # The builder's task signature is ``Callable[[T | None], None]``; wrap
        # _do_set_theme so a None param is a no-op rather than a hard error
        # (matches RelayCommandOf's "param may be missing" contract).
        def _exec(name: str | None) -> None:
            if name is not None:
                self._do_set_theme(name)

        self.set_theme_cmd: RelayCommandOf[str] = RelayCommandOf.builder().task(_exec).build()

    # ── Observable properties ────────────────────────────────────────────

    @property
    def theme(self) -> str:
        return self._theme

    @property
    def mode(self) -> Mode:
        return self._mode

    @property
    def warnings(self) -> tuple[str, ...]:
        return self._warnings

    @property
    def scenario(self) -> ScenarioVM:
        return self._scenario

    @property
    def property_changed(self) -> rx.Observable[str]:
        return self._property_changed_subject

    # ── Commands ────────────────────────────────────────────────────────

    def set_theme(self, name: str) -> None:
        """Direct callable — equivalent to ``set_theme_cmd.execute(name)``."""
        self._do_set_theme(name)

    def _do_set_theme(self, name: str) -> None:
        if name not in _known_themes:
            self._warnings = (*self._warnings, f"Unknown theme: {name}")
            self._property_changed_subject.on_next("warnings")
            self.hub.send(PropertyChangedMessage(self, "AppVM", "Warnings"))
            return
        if name == self._theme:
            return
        self._theme = name
        self._property_changed_subject.on_next("theme")
        self.hub.send(PropertyChangedMessage(self, "AppVM", "Theme"))
        self._save(name)


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------


def make_app_vm(
    *,
    mode: Mode = "native",
    load_theme: Callable[[], str] | None = None,
    persist_theme: Callable[[str], None] | None = None,
    scenario: ScenarioVM | None = None,
    hub: MessageHub[Message] | None = None,
) -> AppVM:
    """Construct an :class:`AppVM`. Mirrors TS ``makeAppVm`` / C# ``AppVMFactory.Create``."""
    return AppVM(
        mode=mode,
        load_theme=load_theme,
        persist_theme=persist_theme,
        scenario=scenario,
        hub=hub,
    )
